### ISSUE TITLE
[ccxt] Exception with ticker

### ISSUE THREAD

**Issue by botum** [gcarq/freqtrade#614]
## Step 2: Describe your environment

  * Python Version: 3.6
  * Branch: feat/objectify-ccxt
  * Last Commit ID: HEAD
 
## Step 3: Describe the problem:

Bot stops aparently because of missing data from Bittrex.

### Steps to reproduce:

  1. dry-run
  2. happens randomly
  
### Observed Results:

  * What happened?
Bot stopped
  * What did you expect to happen?
Bot should keep trying or skip ticker with warning.

### Relevant code exceptions or logs:

  ```
2018-04-20 16:35:47,757 - freqtrade.analyze - WARNING - Outdated history for pair DCT/BTC. Last tick is 7 minutes old
2018-04-20 16:35:47,887 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "./freqtrade/main.py", line 52, in main
    state = freqtrade.worker(old_state=state)
  File "/var/local/user/freqtrade/freqtrade/freqtradebot.py", line 114, in worker
    nb_assets=nb_assets)
  File "/var/local/user/freqtrade/freqtrade/freqtradebot.py", line 126, in _throttle
    result = func(*args, **kwargs)
  File "/var/local/user/freqtrade/freqtrade/freqtradebot.py", line 162, in _process
    state_changed = self.process_maybe_execute_buy()
  File "/var/local/user/freqtrade/freqtrade/freqtradebot.py", line 340, in process_maybe_execute_buy
    if self.create_trade():
  File "/var/local/user/freqtrade/freqtrade/freqtradebot.py", line 285, in create_trade
    (buy, sell) = self.analyze.get_signal(_pair, interval)
  File "/var/local/user/freqtrade/freqtrade/analyze.py", line 127, in get_signal
    ticker_hist = get_ticker_history(pair, interval)
  File "/var/local/user/freqtrade/freqtrade/exchange/__init__.py", line 35, in wrapper
    return f(*args, **kwargs)
  File "/var/local/user/freqtrade/freqtrade/exchange/__init__.py", line 262, in get_ticker_history
    return _API.fetch_ohlcv(pair, timeframe=tick_interval)
  File "/var/local/user/miniconda3/lib/python3.6/site-packages/ccxt/bittrex.py", line 423, in fetch_ohlcv
    return self.parse_ohlcvs(response['result'], market, timeframe, since, limit)
  File "/var/local/user/miniconda3/lib/python3.6/site-packages/ccxt/base/exchange.py", line 971, in parse_ohlcvs
    ohlcv = self.parse_ohlcv(ohlcvs[i], market, timeframe, since, limit)
  File "/var/local/user/miniconda3/lib/python3.6/site-packages/ccxt/bittrex.py", line 403, in parse_ohlcv
    timestamp = self.parse8601(ohlcv['T'] + '+00:00')
KeyError: 'T'

  ```

**Comment by xmatthias** [comment]
According to [this issue](https://github.com/ccxt/ccxt/issues/523) in the ccxt repository, it would appear that this error can happen if the rate-limit is hit (i assume it would also happen if the API does not return correctly).

How many exchanges are in your whitelist? 
how high/low is your `process_throttle_secs` configuration?

Whatever the reason is - if it's only a "one-time" error happening from time to time it should be handled gracefully.

**Comment by xmatthias** [comment]
I had a similar error this morning

```
Traceback (most recent call last):                                                                                                     
  File "/freqtrade/freqtrade/exchange/__init__.py", line 35, in wrapper                                                                
    return f(*args, **kwargs)                                                                                                          
  File "/freqtrade/freqtrade/exchange/__init__.py", line 253, in get_ticker                                                            
    'Could not load tickers due to networking error. Message: {}'.format(e)                                                            
freqtrade.NetworkException: Could not load tickers due to networking error. Message: bittrex https://bittrex.com/api/v1.1/public/getmarketsummary?market=BTC-XRP GET 503 Server Error: Service Temporarily Unavailable for url: https://bittrex.com/api/v1.1/public/getmarketsummary?market=BTC-XRP {"success":true,"message":"","result":[{"MarketName":"BTC-XRP","High":0.00010985,"Low":0.00009770,"Volume":29322483.28568017,"Last":0.00009851,"BaseVolume":3013.04791128,"TimeStamp":"2018-04-21T09:35:36.87","Bid":0.00009838,"Ask":0.00009851,"OpenBuyOrders":3288,"OpenSellOrders":6961,"PrevDay":0.00010250,"Created":"2014-12-22T19:30:27.45"}]}                                         
                                                                                                                                       
During handling of the above exception, another exception occurred:                                                                    
                                                                                                                                       
Traceback (most recent call last):                                                                                                     
  File "/freqtrade/freqtrade/freqtradebot.py", line 158, in _process                                                                   
    state_changed |= self.process_maybe_execute_sell(trade)                                                                            
  File "/freqtrade/freqtrade/freqtradebot.py", line 362, in process_maybe_execute_sell                                                 
    return self.handle_trade(trade)                                                                                                    
  File "/freqtrade/freqtrade/freqtradebot.py", line 374, in handle_trade                                                               
    current_rate = exchange.get_ticker(trade.pair)['bid']                                                                              
  File "/freqtrade/freqtrade/exchange/__init__.py", line 42, in wrapper                               
    return wrapper(*args, **kwargs)                                                                                                    
  File "/freqtrade/freqtrade/exchange/__init__.py", line 42, in wrapper                                                                
    return wrapper(*args, **kwargs)                                                                                                    
  File "/freqtrade/freqtrade/exchange/__init__.py", line 42, in wrapper                                                                
    return wrapper(*args, **kwargs)                                                                                                    
  File "/freqtrade/freqtrade/exchange/__init__.py", line 44, in wrapper                                                                
    raise OperationalException('Giving up retrying: %s', f)                                                                            
freqtrade.OperationalException: ('Giving up retrying: %s', <function get_ticker at 0x7fac629d81e0>)
```

it's not identical, but very similar.
apparently Bittrex returned a 503 for a few tries - resulting in the bot to stop (not crash - but stop). What's odd is that it still appears to contain a (valid) response body - so the data would have been usable.

I guess that is expected behavior since it does not make sense to retry limitless on an API endpoint which is down (or reporting as down, anyway).
However - i think a improved message (more informative than "status: stopped") on Telegram for network retry errors would be helpful to decide if restarting the bot remotely is safe or not.

**Comment by siedi** [comment]
I'm running the code for some weeks (test mode) on several exchanges, getting the same and also similar errors (even 200 returns but empty responses from ccxt). Sometimes ccxt is throwing an error which is caught with an operational error and the bot stops.

The bots stop at least once a day due to those errors. As this is annoying, I'm wondering why the bot just doesn't trigger retries in those cases as well. I changed it locally to also retry for operational errors, and also using the retrying package with their exponential backoff, retrying for 30sec. This helped to have the bots in test mode for 9 days up and running now.

**Comment by botum** [comment]
@siedi right, we've been on this before, there is even a "retrier" function at the exchange app.

Bot should really be sealed against all this, and I don't think stoping the bot is the way, things should go on skiping the pair, or whatever. Uptime is critical for this type of bot.

**Comment by xmatthias** [comment]
@botum skipping the pair assumes that the exchange api is up and running, which may or may not be the case, and i would rather have the bot stop (or at least notify me) instead of it retrying against a "down" exchange for multiple days.

@siedi oddly enough, i've had the bot running in 3 instances for 9 days without modification on binance without it stopping once - then again it stopps almost immediately after restarting.

However, ccxt is now closer to beeing ready, but i don't think it is there yet (there is a calc-error in binance, tracked in #606), as well as multiple stability issues.

**Comment by gitlines** [comment]
I have the feeling this is related:

```
Traceback (most recent call last):
  File "/freqtrade/freqtrade/exchange/__init__.py", line 263, in get_ticker_history
    return _API.fetch_ohlcv(pair, timeframe=tick_interval)
  File "/usr/local/lib/python3.6/site-packages/ccxt/bittrex.py", line 424, in fetch_ohlcv
    raise ExchangeError(self.id + ' returned an empty or unrecognized response: ' + self.json(response))
ccxt.base.errors.ExchangeError: bittrex returned an empty or unrecognized response: {"success":true,"message":"","result":[]}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/freqtrade/freqtrade/freqtradebot.py", line 169, in _process
    state_changed = self.process_maybe_execute_buy()
  File "/freqtrade/freqtrade/freqtradebot.py", line 347, in process_maybe_execute_buy
    if self.create_trade():
  File "/freqtrade/freqtrade/freqtradebot.py", line 292, in create_trade
    (buy, sell) = self.analyze.get_signal(_pair, interval)
  File "/freqtrade/freqtrade/analyze.py", line 112, in get_signal
    ticker_hist = get_ticker_history(pair, interval)
  File "/freqtrade/freqtrade/exchange/__init__.py", line 35, in wrapper
    return f(*args, **kwargs)
  File "/freqtrade/freqtrade/exchange/__init__.py", line 269, in get_ticker_history
    raise OperationalException('Could not fetch ticker data. Msg: {}'.format(e))
freqtrade.OperationalException: Could not fetch ticker data. Msg: bittrex returned an empty
or unrecognized response: {"success":true,"message":"","result":[]}
```
I keep having this issue every day it least once while dry-running the bot on the ccxt-branch with only 1 exchange, bittrex, whitelisted.

@xmatthias my process throttle setting is `"process_throttle_secs": 5`

Does anyone have an idea how to fix this?

**Comment by xmatthias** [comment]
the way others put it ... the api for bittrex is not very stable.
we can't fix the error itself as it's originating from a partial (or invalid) response from the bittrex api endpoint. I don't think it's rate-limiting tbh - the exception should be different in that case - also, ccxt does rate-limiting itself now, so that should be fine i would hope.

ccxt raises the exception, and #619 __should__ fix the bot crashing ... but the data (for that cycle) will miss for that pair.

**Comment by gitlines** [comment]
with the same setup as described above, the bot stopped again, this time without printing the error message to the Telegram Bot (she only said: `Status: Stopping xyz-bot...`).

The docker logs show this message, which clearly gives fault to the Bittrex API which confirms @xmatthias statement about instability of the Bittrex API:

```
2018-04-26 21:48:03,084 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 383, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/local/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/lib/python3.6/ssl.py", line 1009, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 871, in read
    return self._sslobj.read(len, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 357, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 389, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 309, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='bittrex.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 349, in fetch
    proxies=self.proxies
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 521, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='bittrex.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/freqtrade/freqtrade/exchange/__init__.py", line 329, in get_markets
    return _API.fetch_markets()
  File "/usr/local/lib/python3.6/site-packages/ccxt/bittrex.py", line 176, in fetch_markets
    response = self.v2GetMarketsGetMarketSummaries()
  File "/usr/local/lib/python3.6/site-packages/ccxt/bittrex.py", line 682, in request
    response = self.fetch2(path, api, method, params, headers, body)
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 295, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 359, in fetch
    self.raise_error(RequestTimeout, method, url, e)
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 280, in raise_error
    raise exception_type(output)
ccxt.base.errors.RequestTimeout: bittrex GET https://bittrex.com/api/v2.0/pub/Markets/GetMarketSummaries HTTPSConnectionPool(host='bittrex.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/freqtrade/freqtrade/main.py", line 45, in main
    state = freqtrade.worker(old_state=state)
  File "/freqtrade/freqtrade/freqtradebot.py", line 121, in worker
    nb_assets=nb_assets)
  File "/freqtrade/freqtrade/freqtradebot.py", line 133, in _throttle
    result = func(*args, **kwargs)
  File "/freqtrade/freqtrade/freqtradebot.py", line 153, in _process
    ) if nb_assets else self.config['exchange']['pair_whitelist']
  File "/freqtrade/freqtrade/freqtradebot.py", line 224, in _refresh_whitelist
    markets = exchange.get_markets()
  File "/freqtrade/freqtrade/exchange/__init__.py", line 332, in get_markets
    'Could not load markets due to networking error. Message: {}'.format(e)
freqtrade.NetworkException: Could not load markets due to networking error. Message: bittrex GET https://bittrex.com/api/v2.0/pub/Markets/GetMarketSummaries HTTPSConnectionPool(host='bittrex.com', port=443): Read timed out. (read timeout=10)
2018-04-26 21:48:03,086 - freqtrade.rpc.rpc_manager - INFO - *Status:* `Stopping trader...`
2018-04-26 21:48:03,169 - freqtrade.freqtradebot - INFO - Stopping trader and cleaning up modules...
2018-04-26 21:48:03,170 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc.telegram ...
```

I restarted with Binance now and it works fine since 4 hours. However, I believe I saw the error described in my previous post first using Binance when I tried the ccxt-branch for the first time a week ago...

**Comment by gitlines** [comment]
Alright, it happens on Binance as well, just less frequently:

```
2018-05-02 06:26:01,428 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 383, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/local/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/lib/python3.6/ssl.py", line 1009, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 871, in read
    return self._sslobj.read(len, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 357, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 389, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 309, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 349, in fetch
    proxies=self.proxies
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 521, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/freqtrade/freqtrade/exchange/__init__.py", line 329, in get_markets
    return _API.fetch_markets()
  File "/usr/local/lib/python3.6/site-packages/ccxt/binance.py", line 350, in fetch_markets
    response = self.publicGetExchangeInfo()
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 298, in request
    return self.fetch2(path, api, method, params, headers, body)
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 295, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 359, in fetch
    self.raise_error(RequestTimeout, method, url, e)
  File "/usr/local/lib/python3.6/site-packages/ccxt/base/exchange.py", line 280, in raise_error
    raise exception_type(output)
ccxt.base.errors.RequestTimeout: binance GET https://api.binance.com/api/v1/exchangeInfo HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/freqtrade/freqtrade/main.py", line 45, in main
    state = freqtrade.worker(old_state=state)
  File "/freqtrade/freqtrade/freqtradebot.py", line 121, in worker
    nb_assets=nb_assets)
  File "/freqtrade/freqtrade/freqtradebot.py", line 133, in _throttle
    result = func(*args, **kwargs)
  File "/freqtrade/freqtrade/freqtradebot.py", line 153, in _process
    ) if nb_assets else self.config['exchange']['pair_whitelist']
  File "/freqtrade/freqtrade/freqtradebot.py", line 224, in _refresh_whitelist
    markets = exchange.get_markets()
  File "/freqtrade/freqtrade/exchange/__init__.py", line 332, in get_markets
    'Could not load markets due to networking error. Message: {}'.format(e)
freqtrade.NetworkException: Could not load markets due to networking error. Message: binance GET https://api.binance.com/api/v1/exchangeInfo HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=10)
2018-05-02 06:26:01,431 - freqtrade.rpc.rpc_manager - INFO - *Status:* `Stopping trader...`
2018-05-02 06:26:01,534 - freqtrade.freqtradebot - INFO - Stopping trader and cleaning up modules...
2018-05-02 06:26:01,534 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc.telegram ...
``
`
Is there any way to somehow auto restart the bot - or better the docker container in which it is running? Because in the two cases above, the whole program crashes, docker stops the container and the error message is printed by the corresponding telegram bot. 

This scenario is by the in contrast to my [first post in this issue](https://github.com/gcarq/freqtrade/issues/614#issuecomment-384693979) where the bot stops, prints "stopping trader..." via the telegram bot and can be started via telegram again.

@xmatthias , since you `internals.process_throttle_secs` earlier, I wonder what exactly is being throttled by this setting? Does it slow down the rate of API requests? This would be my intuitive interpretation for setting the process throttle; however, I though rate limiting is taking care of this already.

**Comment by xmatthias** [comment]
think of the bot as running in "cycles" ... 
each cycle
* checks all pairs for possible trades  
* checks all trades for possible sells

the `process_throttle_secs` simply defines a "sleep" time between 2cycles - it will implicitly slow down API calls (as it does nothing for X seconds).

code for this is all [close together](https://github.com/gcarq/freqtrade/blob/develop/freqtrade/freqtradebot.py#L112-L138).

**Comment by gcarq** [comment]
Should be fixed on `feat/objectify-ccxt` at a5c1547251b5ff0ba10153ff26f639e17281be3a
